### PR TITLE
feat(PLM-1741) Support Nested Spec Class

### PR DESCRIPTION
**Pull request by kilikkuo**
[Why]
It could be eaiser for maintenance, can be easier to identify that the class is used by some other else

[How]
1. Revise the extract fields logic, explicitly check for Checker instance
2. Add test cases & simplify old cases

[PLM-1741]

Reason
--------------

Changes
--------------

Test Scope
--------------

Checks
--------------
- [x] Unit tests are included or not applicable.

[PLM-1741]: https://hardcoretech.atlassian.net/browse/PLM-1741?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: data_spec_validator/spec/validators.py [hunk 1]
```diff
@@ -349,11 +345,21 @@ def validate(value, extra, data) -> Tuple[bool, Union[Exception, str]]:
...
 
+    @staticmethod
+    def _extract_fields(spec) -> List[str]:
+        raise_if(type(spec) != type, RuntimeError(f'{spec} should be just a class'))
+
+        def _pick_checker_field(acc, item):
+            if isinstance(item[1], Checker):
+                acc.append(item[0])
+            return acc
+        return list(functools.reduce(_pick_checker_field, spec.__dict__.items(), []))
```
Comment (CJHwong) [inline data_spec_validator/spec/validators.py]: nit:
reduce, filter are kind of hard to understand...
```python

```

```suggestion
        return [
            f_name 
            for f_name, checker in spec.__dict__.items()
            if isinstance(checker, Checker)
        ]
```

File: test/test_nested_spec.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        assert validate_data_spec(ok_data, NestedSpec)
+
+        nok_data = dict(
+            c1_f=3,
+            int_f='9',
+            str_f=3,
+            c2_f=None,
+            c3_f=[],
+        )
+        assert is_something_error(TypeError, validate_data_spec, nok_data, NestedSpec)
```
Comment (CJHwong) [inline test/test_nested_spec.py:63]: nit:

Flatten the nested checkers is a clean move but it makes the error message a little implicit.

For example, when setting `nok_data['c1_f']` as `dict(f_1=3.4, d_1=dict(a=1), b_1=True, s_1=None)`

I'd expect seeing the error message have `field: NestedSpec.c1_f.s_1` but it shows `field: ChildSpec1.s_1` instead.

Maybe we can use some metadata in the validator for constructing verbose error messages?

**Review by CJHwong — APPROVED** [review]

File: data_spec_validator/spec/validators.py (hunk 1 above)
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py]: nice!

File: test/test_nested_spec.py (hunk 2 above)
Comment (kilikkuo) [reply test/test_nested_spec.py:63]: good idea, followup created, https://hardcoretech.atlassian.net/browse/PLM-1799

**Pull request merged**